### PR TITLE
Create JuliaFormatter GitHub Action

### PR DESCRIPTION
**Pull request by simonbyrne**
# Description

On each PR, this will run the JuliaFormatter.jl on all modified .jl files, failing if the formatting modifies any of the files.

I originally tried using https://github.com/julia-actions/julia-format, but it wasn't really sufficient for our needs, so it made more sense to use a custom workflow.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [x] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions
- [x] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by jkozdon** [comment]
Can we add something to the `README.md` on how to apply this locally prior to doing a PR?

**Comment by simonbyrne** [comment]
Yes, though it should probably go in the developer docs / CONTRIBUTING.md (#794).

**Comment by simonbyrne** [comment]
Alright, this is working now. To apply the formatter locally, call
```
julia .dev/format.jl
```
from the main CLIMA directory

**Comment by charleskawczynski** [comment]
This command applies the formatter to the diff of our contribution, correct?

**Comment by simonbyrne** [comment]
It applies it to all files in the diff (so even if you just touch part of a file, it will apply it to all of it)

**Comment by jkozdon** [comment]
> It applies it to all files in the diff (so even if you just touch part of a file, it will apply it to all of it)

Is the plan to do the repo in one fell swoop or just the files that PRs touch as they come in?

**Comment by simonbyrne** [comment]
What we settled on was to do files as changes come in, then do the rest once we hit a reasonable threshold.

**Comment by leios** [comment]
I have been playing around with this on my own fork by setting the sb/julia-formatter branch as default. I cannot seem to get "check formatting diff" to pass without error, but I also cannot quite figure out what it is diffing against with `git diff --color=always --exit-code` in JuliaFormatter.yml

[Here is the PR I am looking at](https://github.com/leios/CLIMA/pull/2). I could be doing something wrong.

In that fork, I set the format.jl file to look for changes from the sb/julia-formatter branch and running the formatter locally seems to work; however, the indenting to four spaces seems to be a bit weird in some cases.

Running `git diff --color=always --exit-code` after using the formatter locally looks like this:

```
diff --git a/file.jl b/file.jl
index d79be543..926d905e 100644
--- a/file.jl
+++ b/file.jl
@@ -2,8 +2,8 @@
 
 
 function this_wont_work()
- 1
-  2
-   3
+    1
+    2
+    3
     4
 end
```
(I promise it's colored on my screen)

The diff is 1400 lines on the fork, though, and I cannot quite find the error. I'll keep poking at it a bit more, just wanted to update everyone.

**Comment by simonbyrne** [comment]
The diff is comparing what is in your PR vs what it would get if it ran the formatter against the .jl files changed in your PR. So that output is correct, since the formatter modified those lines of the files.

For some reason it thinks CLIMA.jl and src/Mesh/Interpolations.jl have changed:
https://github.com/leios/CLIMA/pull/2/checks?check_run_id=488246829#step:5:33
but I'm not sure why...

**Comment by simonbyrne** [comment]
Ah, it's because those files have been modified in your master, but not in your `bad_code` branch.

**Comment by simonbyrne** [comment]
Ah, we need to use `...` to figure out the common ancestor when doing the diffs: https://stackoverflow.com/a/38436974/392585

**Comment by simonbyrne** [comment]
It works: https://github.com/climate-machine/CLIMA/pull/797/checks?check_run_id=488444983#step:6:14

**Comment by leios** [comment]
Right, it's still not building on my PR, but I guess if it works here that is all that matters. The diff is correct, but it's giving an error still. 

Apparently the margin rule does not apply to comments. If you have a comment exceeding 80 columns, it is not automatically wrapped. I cannot seem to find this functionality in the JuliaFormatter docs, so it might be overlooked? If so, I'll create an issue there.

**Comment by simonbyrne** [comment]
It seems to be "working" correctly now, in that the `git diff` check fails because your code isn't correctly formatted: https://github.com/leios/CLIMA/pull/2/checks?check_run_id=488498421

**Comment by leios** [comment]
Ah, I see. It's catching to make sure the format is good, but we still run the format command locally. That's fine.

I cannot seem to get the formatter to fix comments over the margin, nor can I find anything about it in JuliaFormatter, so I'll create an issue there.

**Comment by simonbyrne** [comment]
Notes to add to docs:
- how to apply
- make formatting changes separate commit

**Comment by simonbyrne** [comment]
I added it to CONTRIBUTING.md.

@leios can you review the text, and merge when you're happy?

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+```
+
+Formatting changes should be done as a separate commit, ideally the last commit of the pull request (you may want to leave it until all other changes have been approved).
+
+## Bors and CI
+
+All commits that end up in the CLIMA repository must pass Continuous Integration (CI).
+When a PR is updated, it will automatically be run on Microsoft Azure for Linux, Windows, and MacOS; however, because CLIMA is heterogeneous and must also run on GPU hardware, we also manually launch CI with Bors.
+
+To test to see if all bors CI will pass, please type `bors test` in a separate comment in the PR.
```
Comment (kpamnany) [inline CONTRIBUTING.md]: This should be `bors try`.

Comment (simonbyrne) [reply CONTRIBUTING.md]: @leios what's the status of this?

Comment (leios) [reply CONTRIBUTING.md]: Kiran and I were having some discussions about what to put into it and who the audience should be. I'll push a new version in ~30 minutes, just finishing up some KA stuff.

**Comment by kpamnany** [comment]
Gentle reminder to squash first. :)

**Pull request merged**